tcp-native/src/main/java/org/ballerinalang/stdlib/tcp/TcpClient.java

```diff
@@ -119,21 +118,21 @@ private void setSSLHandler(SocketChannel channel, BMap<BString, Object> secureSo
         channel.pipeline().addLast(Constants.SSL_HANDSHAKE_HANDLER, new SslHandshakeEventHandler(tcpClientHandler));
     }
 
-    public void writeData(byte[] bytes, Future callback) throws InterruptedException {
+    public void writeData(byte[] bytes, Future callback) {
         if (channel.isActive()) {
-            channel.writeAndFlush(Unpooled.wrappedBuffer(bytes)).addListener((ChannelFutureListener) future -> {
-                if (future.isSuccess()) {
-                    callback.complete(null);
-                } else {
-                    callback.complete(Utils.createSocketError("Failed to send data " + future.cause().getMessage()));
-                }
-            });
+            WriteCallbackService writeCallbackService = new WriteCallbackService(Unpooled.wrappedBuffer(bytes),
+                    callback, channel);
+            writeCallbackService.writeData();
+            if (!writeCallbackService.isWriteCalledForData()) {
```

Comment:
I think we need to do the below before doing writingdata

Reply:
fixed